File: frontend/src/types/tn.ts
export interface TnEntry {
  id: number
  name: string
  family: string
  tn_group: string
  synonyms: string | null
  isoform: string | null
  accession_number: string | null

  origin: string | null
  mge_type: string | null
  related_elements: string | null

  length: number | null
  ir: string | null
  dr: number | null
  orf: string | null
  irl: string | null
  irr: string | null
  left_flank: string | null
  right_flank: string | null

  transposition: string | null
  direct_repeat: string | null

  dna_sequence: string | null

  orf1_name: string | null
  orf1_length: number | null
  orf1_begin: number | null
  orf1_end: number | null
  orf1_strand: string | null
  orf1_fusion_orf: string | null
  orf1_function: string | null
  orf1_chemistry: string | null
  orf1_sequence: string | null

  orf2_name: string | null
  orf2_length: number | null
  orf2_begin: number | null
  orf2_end: number | null
  orf2_strand: string | null
  orf2_fusion_orf: string | null
  orf2_function: string | null
  orf2_chemistry: string | null
  orf2_sequence: string | null

  status: string
  submitted_by: number | null
  reviewed_by: number | null
  reviewed_at: string | null
  created_at: string
  updated_at: string
}

export interface TnFilter {
  keyword?: string
  family?: string
  tn_group?: string
  origin?: string
  mge_type?: string
  status?: string
  page: number
  page_size: number
}

export interface TnListResponse {
  items: TnEntry[]
  total: number
  page: number
  page_size: number
}

File: backend/app/schemas/tn.py
from pydantic import BaseModel
from datetime import datetime
from typing import Optional, List


class TnEntryBase(BaseModel):
    name: str
    family: str
    tn_group: str
    synonyms: Optional[str] = None
    isoform: Optional[str] = None
    accession_number: Optional[str] = None

    origin: Optional[str] = None
    mge_type: Optional[str] = None
    related_elements: Optional[str] = None

    length: Optional[int] = None
    ir: Optional[str] = None
    dr: Optional[int] = None
    orf: Optional[str] = None
    irl: Optional[str] = None
    irr: Optional[str] = None
    left_flank: Optional[str] = None
    right_flank: Optional[str] = None

    transposition: Optional[str] = None
    direct_repeat: Optional[str] = None

    dna_sequence: Optional[str] = None

    orf1_name: Optional[str] = None
    orf1_length: Optional[int] = None
    orf1_begin: Optional[int] = None
    orf1_end: Optional[int] = None
    orf1_strand: Optional[str] = None
    orf1_fusion_orf: Optional[str] = None
    orf1_function: Optional[str] = None
    orf1_chemistry: Optional[str] = None
    orf1_sequence: Optional[str] = None

    orf2_name: Optional[str] = None
    orf2_length: Optional[int] = None
    orf2_begin: Optional[int] = None
    orf2_end: Optional[int] = None
    orf2_strand: Optional[str] = None
    orf2_fusion_orf: Optional[str] = None
    orf2_function: Optional[str] = None
    orf2_chemistry: Optional[str] = None
    orf2_sequence: Optional[str] = None


class TnEntryCreate(TnEntryBase):
    pass


class TnEntryUpdate(BaseModel):
    name: Optional[str] = None
    family: Optional[str] = None
    tn_group: Optional[str] = None
    synonyms: Optional[str] = None
    isoform: Optional[str] = None
    accession_number: Optional[str] = None
    origin: Optional[str] = None
    mge_type: Optional[str] = None
    related_elements: Optional[str] = None
    length: Optional[int] = None
    ir: Optional[str] = None
    dr: Optional[int] = None
    orf: Optional[str] = None
    irl: Optional[str] = None
    irr: Optional[str] = None
    left_flank: Optional[str] = None
    right_flank: Optional[str] = None
    transposition: Optional[str] = None
    direct_repeat: Optional[str] = None
    dna_sequence: Optional[str] = None
    orf1_name: Optional[str] = None
    orf1_length: Optional[int] = None
    orf1_begin: Optional[int] = None
    orf1_end: Optional[int] = None
    orf1_strand: Optional[str] = None
    orf1_fusion_orf: Optional[str] = None
    orf1_function: Optional[str] = None
    orf1_chemistry: Optional[str] = None
    orf1_sequence: Optional[str] = None
    orf2_name: Optional[str] = None
    orf2_length: Optional[int] = None
    orf2_begin: Optional[int] = None
    orf2_end: Optional[int] = None
    orf2_strand: Optional[str] = None
    orf2_fusion_orf: Optional[str] = None
    orf2_function: Optional[str] = None
    orf2_chemistry: Optional[str] = None
    orf2_sequence: Optional[str] = None


class TnEntryResponse(TnEntryBase):
    id: int
    status: str
    submitted_by: Optional[int] = None
    reviewed_by: Optional[int] = None
    reviewed_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class TnFilter(BaseModel):
    keyword: Optional[str] = None
    family: Optional[str] = None
    tn_group: Optional[str] = None
    origin: Optional[str] = None
    mge_type: Optional[str] = None
    status: Optional[str] = None
    page: int = 1
    page_size: int = 20


class TnListResponse(BaseModel):
    items: List[TnEntryResponse]
    total: int
    page: int
    page_size: int
